docs/markdown/Using Pants/concepts/targets.md

```diff
@@ -388,3 +388,238 @@ shunit2_tests(
+# The following is equivalent:
+__dependencies_rules__(
+  <ruleset-nested-1>,
+  <ruleset-nested-2>,
+  <ruleset-top-1>,
+  <ruleset-top-2>,
+)
+# with:
+__dependencies_rules__(
+  <ruleset-nested-1>,
+  <ruleset-nested-2>,
+  extend=True,
+)
+```
+
+
+Rule sets
+---------
+
+As there may be many targets and files with dependencies, odds are that they won't all share the same set of rules. The rules keywords accepts multiple sets of rules, or "rule sets", along with "selector rules" that is used to select which set to use for each target.
+
+The overall structure is (example with 2 rule sets):
+
+```python
+# BUILD
+
+__dependencies_rules__(
+
+  # Rule set 1
+  (<selectors>, <rule a>, <rule b>, ...),
+
+  # Rule set 2
+  (<selectors>, <rules>, ...),
+
+  ...
+)
+```
+
+The rules are just string values using a [glob syntax](doc:targets#glob-syntax) for pattern matching and may be grouped together for readability (how rules are grouped does not affect how they are applied). The selector is a dictionary value with properties describing what targets its associated rules apply to and together this pair of selector(s) and rules is called a __Rule Set__. A rule set may have multiple selectors wrapped in a list/tuple.
+
+The selector has three properties: `type`, `tags` and `path`. From the above example, when determining which rule set to apply for the dependencies of `src/a/main.py` Pants will look for the first selector for `src/a/BUILD` that satisifies the properties `type=python_sources`, `tags=["apps"]` and `path=src/a/main.py`. The selection is based on exclusion so only when there is a property value and it doesn't match the targets property it will move on to the next selector, so the lack of a property will be considered to match anything. Consequently an empty selector matches all targets.
+
+The values of a selector supports wildcard patterns (or globs) in order to have a single selector match multiple different targets. The `path` property uses the same [glob syntax](doc:targets#glob-syntax) as the rules, while `type` and `tags` use a simpler one described below. When listing multiple values for the `tags` property, the target must have all of them in order to match. Spread the tags over multiple selectors in order to switch from AND to OR as required. The target `type` to match against will be that of the type used in the BUILD file, as the path (and target address) may refer to a generated target it is the target generators type that will be used during the selector matching process.
+
+The simpler glob syntax used by the `type` and `tags` selector values supports `*` as a match anything and is otherwise case sensitive. (implementation detail: it relies on the `fnmatch` python library so there is a bit more syntax available, but if you find yourself using more than `*` please let us know in case we switch to something else).
+
+### NOTE/TODO: maybe drop the use of `fnmatch` entirely, as supporting just `*` is real easy.
```

Comment:
I'm in favour of dropping the remaining use of `fnmatch` to limit the initial glob syntax supported. (see the implementation note in the previous paragraph for primary motivation for this)

> (implementation detail: it relies on the `fnmatch` python library so there is a bit more syntax available, but if you find yourself using more than `*` please let us know in case we switch to something else).